runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaInputMetricOp.java

```diff
@@ -0,0 +1,89 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.samza.runtime;
+
+import java.math.BigInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.beam.runners.samza.metrics.BeamTransformMetricRegistry;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * SamzaInputMetricOp is a {@link SamzaMetricOp} that emits & maintains default transform metrics
+ * for input PCollection to the transform. It emits the input throughput and maintains avg arrival
+ * time for input PCollection per watermark.
+ *
+ * <p>Assumes that {@code SamzaInputMetricOp#processWatermark(Instant, OpEmitter)} is exclusive of
+ * {@code SamzaInputMetricOp#processElement(Instant, OpEmitter)}. Specifically, the processWatermark
+ * method assumes that no calls to processElement will be made during its execution, and vice versa.
+ *
+ * @param <T> The type of the elements in the input PCollection.
+ */
+public class SamzaInputMetricOp<T> extends SamzaMetricOp<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(SamzaInputMetricOp.class);
+  // Counters to maintain avg arrival time per watermark for input PCollection.
+  private AtomicLong count;
+  private AtomicReference<BigInteger> sumOfTimestamps;
+
+  public SamzaInputMetricOp(
+      String pValue,
+      String transformFullName,
+      BeamTransformMetricRegistry beamTransformMetricRegistry) {
+    super(pValue, transformFullName, beamTransformMetricRegistry);
+    this.count = new AtomicLong(0L);
+    this.sumOfTimestamps = new AtomicReference<>(BigInteger.ZERO);
+  }
+
+  @Override
+  public void processElement(WindowedValue<T> inputElement, OpEmitter<T> emitter) {
+    count.incrementAndGet();
+    sumOfTimestamps.updateAndGet(sum -> sum.add(BigInteger.valueOf(System.nanoTime())));
+    beamTransformMetricRegistry
+        .getTransformMetrics()
+        .getTransformInputThroughput(transformFullName)
+        .inc();
+    emitter.emitElement(inputElement);
+  }
+
+  @Override
+  public void processWatermark(Instant watermark, OpEmitter<T> emitter) {
+    if (LOG.isDebugEnabled()) {
```

Comment:
We only need this line when the errMsg is expensive to construct no?

Reply:
I dont understand this, we need this in issues we need to trace code using watermark / task combination since they are vary across task

Reply:
I thought we only need the `if (LOG.isDebugEnabled()) {` check when the construction of the log message is expensive. Else, we can just write `LOG.debug(...)`, and let log4j filter out the message.

Reply:
Oh i see what you mean, i proactively added it right now since processWatermark itself is not concurrent right now, we do timer firing in processWatermark for event based watermarks